### PR TITLE
[BEAM-10652] Allow Clustering without Partition in BigQuery

### PR DESCRIPTION
**Pull request by brucearctor**
Allow Clustering without Partition in BigQuery

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Follow up from https://github.com/apache/beam/pull/15548

**Comment by brucearctor** [comment]
@TheNeuralBit -- can you have a look to ensure I'm not way off base with what I'm doing here?  I made a new PR here.  I think this addresses the current issue as written.  Clustering does not appear to work by itself (without TimePartitioning) in all cases, but this does work for the cases that were added.  Overall an improvement, but I should I probably add some docs/writeup somewhere for these cases that do work ( to acknowledge is just a subset ).  Naturally, will squash all these commits :-)

**Comment by aaltay** [comment]
@kerrydc @chamikaramj - This looks like a BQ change. Could you suggest a potential reviewer for this PR?

**Comment by TheNeuralBit** [comment]
I'd feel better if @chamikaramj reviewed this since he has more context on the original reasoning for disallowing clustering without partitioning. My involvement here was just filing BEAM-10652 after it came up in a customer issue.

**Comment by brucearctor** [comment]
Great -- @chamikaramj do take a look if you're up for it or please advise if I should pull in someone else ( I am a committer, and probably being overly cautious, I can also see about pulling anyone else in if you're busy ) -- even though lots of commits, I think the changes are rather straightforward [ once squashed ].

**Comment by chamikaramj** [comment]
Ack. Will check.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:105]: Do you need to set this coder manually ? I thought we choose the appropriate coder when clustering is enabled.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:151]: Can we also verify the data written ?

**Review by chamikaramj — COMMENTED** [review]
Thanks.

**Comment by brucearctor** [comment]
Java Run PostCommit

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java
Comment (brucearctor) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:151]: Sure ...  I had set a default table expiration of 5 days, to not unnecessarily store IT data.  Currently that is 5 days -- do we think we should have a longer retention period?

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
run java postcommit

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java
Comment (brucearctor) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:105]: Easiest way to verify ( without actually thinking much :-) ) is going to be removing these lines and see if things still work.  Once I re-confirm the below (that also writing the data), I'll then try removing these lines to verify whether works without.  

Off handedly, I think it is needed ( I think we default to v2 coder, and need an override ... but ... is great if this is already handled and setting manually not required ).  If this turns out to be necessary, then that seems like a new issue which I can at least file a ticket - and can consider addressing after this.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java
Comment (brucearctor) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:151]: The data is being written/created (verified manually).    Do you have a suggestion for an Assert to use ( check row count,?check a specific value? ).  Clustering *should* only effect the table being created and not the data itself, but doesn't hurt to add some additional checks/tests/assertions.  

Perhaps there is an example from another IT that you think would be good?

Comment (chamikaramj) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:151]: Yeah, I mean to check the properties of the data (for example, rowcount, a checksum of the data) in the same IT using an assert. There's an example here: https://github.com/apache/beam/blob/419c03c71d47bc1a4001735e260fcfe083da164d/examples/java/src/test/java/org/apache/beam/examples/cookbook/BigQueryTornadoesIT.java#L70

Comment (brucearctor) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:151]: ya, that makes sense to have.  Rowcount super straightforward.  Hash is better, but ... I don't see a great way to easily generate to put into the test ( get it the first time ).  I'll think more on the approach today.  Thanks!

**Comment by brucearctor** [comment]
run java postcommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
BTW, also please make sure that this works for all BQ write methods: https://github.com/apache/beam/blob/7fa5387ffac4f2801077f2e55aa2eba7a47036d9/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L1728

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
You can just try running the same pipeline with three write methods (no need to add additional tests). For testing streaming, for Dataflow, just add the "--streaming" flag. The pipeline will end since the input is bounded but it still runs using the streaming runner.

**Comment by brucearctor** [comment]
@chamikaramj  -- any idea why `Run Java PreCommit` and others do not seem to trigger those testing suites?  Is this generally flaky, or am I not doing something that would otherwise be helpful?

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
I think Beam Jenkins is down currently: https://issues.apache.org/jira/browse/INFRA-22878

**Comment by brucearctor** [comment]
> I think Beam Jenkins is down currently: https://issues.apache.org/jira/browse/INFRA-22878

That seems to explain it, thanks!

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Spotless PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java_Examples_Dataflow_Java11 PreCommit

**Comment by brucearctor** [comment]
Run Java_Examples_Dataflow_Java17 PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java
Comment (brucearctor) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:105]: It looks like this is needed -- from the 'test' ( by removing setting this manually ) the IT failed, but it succeeds if it is in there.  

Am inclined to mark this as resolved, but @chamikaramj do advise.

Comment (chamikaramj) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:105]: Sg. Resolving.

**Comment by brucearctor** [comment]
Run Spotless PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java
Comment (brucearctor) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryClusteringIT.java:154]: I implimented # of lines and expected bytes -- This seemed sufficient and was easier to implement than using bigquerymatcher.

**Comment by brucearctor** [comment]
> You can just try running the same pipeline with three write methods (no need to add additional tests). For testing streaming, for Dataflow, just add the "--streaming" flag. The pipeline will end since the input is bounded but it still runs using the streaming runner.

@chamikaramj -- this is hurting my head for some reason ...  Where would you suggest the pipelines run to test these things?  I've only been thinking in terms of IntegrationTests, is there another way you'd suggest to test this ( as it sounds like you think a one-off is sufficient, as a sanity check rather than an ongoing testing need ).  Thanks!

**Comment by chamikaramj** [comment]
If it's hard to run the Jenkins integration test for all write methods (I don't have a command to do so unfortunately) you can just try running a simple pipeline that write to BQ (for example [1]) to make sure the feature works for various write methods.

If it makes easier you can just operate on top of a Maven project created using the archetype given at [2].

If you want to push code from the Beam Gradle repo to local Maven repo you can use the command at [3].

[1] https://github.com/apache/beam/blob/master/examples/java/src/main/java/org/apache/beam/examples/cookbook/BigQueryTornadoes.java
[2] https://beam.apache.org/get-started/quickstart-java/
[3] https://cwiki.apache.org/confluence/display/BEAM/Gradle+Tips#GradleTips-PublishamodulesnapshotreleasetomavenLocal

**Comment by brucearctor** [comment]
Thanks, Cham.  Even if hard, doesn't mean not worth doing.  **My philosophy is that if it is important to test once before merging then should be part of ongoing tests....**, so will at least try that.  

Though, it doesn't seem to make sense to test the STORAGE_API_AT_LEAST_ONCE , since the test might fail ( if configured to expect exactly-once ), even though the API would be working as 'expected'/advertised.

**Comment by brucearctor** [comment]
and, I think I found the correct 'withMethod' for more tests in Jenkins!

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
[](https://github.com/brucearctor)Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Spotless PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Spotless PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
Run Java PreCommit

**Comment by brucearctor** [comment]
Run Java PostCommit

**Comment by brucearctor** [comment]
@chamikaramj  -- how's this code look?  This is an improvement, that doesn't break any existing tests -- so, I think is OK to merge.

Clustering is not created properly for STORAGE_API_AT_LEAST_ONCE or STORAGE_WRITE_API -- so those will be new tickets that I add.  Again, since this is an improvement over existing functionality ( and does not break things ), it seems OK to merge.  I might revisit to get those working later, but for now the tests are for 'DEFAULT', and 'FILE_LOADS'.  

I will write some follow up tickets for:
a) document this functionality.  at least in https://beam.apache.org/documentation/io/built-in/google-bigquery/ , but perhaps anywhere else suggested
b) get clustering working with STORAGE_API_AT_LEAST_ONCE
c) get clustering working with STORAGE_WRITE_API
...?  Anything else top of mind ( what else would we want to see, but needn't be captured by this current improvement?  


Also, would be curious to learn the implementation differences between STORAGE_API_AT_LEAST_ONCE and STORAGE_WRITE_API ... I wonder if solved for one, if that fixes both.  


I see some related issues that I'd like to see addressed ahead of implementing any 'fix' or additional integration tests just for clustering, when using the storage write api --> ex:
https://issues.apache.org/jira/browse/BEAM-13747
and
https://issues.apache.org/jira/browse/BEAM-13753 / https://issues.apache.org/jira/browse/BEAM-13990


If you agree is fine, do squash/merge, or give me the LGTM.  Thanks!

**Comment by brucearctor** [comment]
@chamikaramj  -- LGTY?

**Review by chamikaramj — APPROVED** [review]
Thanks.

**Pull request merged**
